exporter/exporterhelper/internal/sized_elements_channel.go

```diff
@@ -0,0 +1,111 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Int64
+
+	// We need to store the capacity in a separate field because the capacity of the channel can be higher.
+	// It happens when we restore a persistent queue from a disk that is bigger than the pre-configured capacity.
+	cap int64
+	ch  chan T
+
+	// The following fields are used by the persistent queue to initialize the queue with the elements recovered from
+	// the disk. They have to be tracked separately because the actual total size of all the restored elements is not
+	// guaranteed to be equal initTotalSize.
+	initTotalSize     int64
+	initElementsCount int64
+	initSizeMutex     sync.Mutex
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int) *sizedElementsChannel[T] {
+	return &sizedElementsChannel[T]{
+		used: &atomic.Int64{},
+		cap:  int64(capacity),
+		ch:   make(chan T, capacity),
+	}
+}
+
+// enqueue puts the element into the queue with the given sized if there is enough capacity.
+// Returns an error if the queue is full. The callback is called before the element is committed to the queue.
+// If the callback returns an error, the element is not put into the queue and the error is returned.
+func (vcq *sizedElementsChannel[T]) enqueue(el T, size int64, callback func() error) error {
+	if vcq.initTotalSize+vcq.used.Add(size) > vcq.cap {
```

Comment:
Wrong usage of `initTotalSize` since is not protected by the mutex.

Reply:
Ok. I added RW mutex to handle it

Reply:
This slows down the hot path a bit, but not significantly. I updated the benchmark results